File: Olympiad/POI/official/06-XIII/okr/prog/okrs0.cpp
/*************************************************************************
 *                                                                       *
 *                   XIII Olimpiada Informatyczna                        *
 *                                                                       *
 *   Rozwi-zanie zadania: OKR (Okresy s³ów)                              *
 *   Plik:                okrs0.cpp                                      *
 *   Autor:               Karol Cwalina                                  *
 *   Opis:                Znajduje dlugosc pasujacego na danej pozycji   *
 *                        prefiksu, czas O(n^2)                          *
 *                                                                       *
 *************************************************************************/

#include <cstdio>

using namespace std;

int main (void) {
    int n;
    scanf("%d", &n);
    char *tekst = new char[n+1];
    scanf("%s", tekst);

    int *pref = new int[n];

    /* Dla kazdej pozycji szukamy dlugosci prefiksu
    tekstu, ktory zaczyna sie na tej pozycji */
    pref[0]=0;
    for (int i=1; i<n; pref[++i]=0) {
        int j;
        for (j=0; tekst[i+j] == tekst[j]; ++j) ;
        pref[i]=j;
    }
    /* Nastepnie dla kazdej pozycji szukamy najblizszej,
    z ktorej wychodzi prefiks obejmujacy nasza pozycje */
    long long odp=0;
    int max_pref=0;
    for (int i=1; i<n; ++i) {
        /* Najpierw sprawdzamy, czy w ogole istnieje jakis prefiks */
        if (i+pref[i]-1 > max_pref)
            max_pref = i+pref[i]-1;
        if (max_pref < i)
            continue;
        /* A dopiero pozniej szukamy maksymalnego */
        for (int j=i; 2*j >= i; --j)
            if (j+pref[j] > i) {
               odp += j;
               break;
            }
    }

    printf("%lld\n", odp);

    return 0;
}

File: Olympiad/POI/official/06-XIII/okr/prog/okrs0.pas
(*************************************************************************
 *                                                                       *
 *                   XIII Olimpiada Informatyczna                        *
 *                                                                       *
 *   Rozwi-zanie zadania: OKR (Okresy s³ów)                              *
 *   Plik:                okrs0.pas                                      *
 *   Autor:               Karol Cwalina                                  *
 *   Opis:                Znajduje dlugosc pasujacego na danej pozycji   *
 *                        prefiksu, czas O(n^2)                          *
 *                                                                       *
 *************************************************************************)

program okresy_slow;
const MaxN=1000000;
var i,j,max_pref,n : LongInt;
    odp : Int64;
    tekst : array [1..MaxN+1] of Char;
    pref : array [1..MaxN] of LongInt;

begin
    Readln(n);
    for i := 1 to n do Read(tekst[i]);
    tekst[n+1] := '$';

    pref[1] := 0;
    for i := 2 to n do begin
        j := 0;
        while tekst[i+j] = tekst[j+1] do Inc(j);
        pref[i] := j;
    end;

    odp := 0; max_pref := 0;
    for i := 2 to n do begin
        if i+pref[i]-1 > max_pref then max_pref := i+pref[i]-1;
        if max_pref < i then continue;
        for j := i downto (i+3) div 2 do
            if j+pref[j] > i then begin
                odp := odp+j-1;
                break;
            end;
    end;

    Writeln (odp);
end.
